WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -9,46 +9,176 @@ import Alamofire
 ///
 /// API doc: https://developer.wordpress.com/docs/oauth2/
 struct WordPressDotComAuthenticator {
-    enum Error: Swift.Error {
+    enum SignInContext {
+        // Sign in to the app with a WordPress.com account.
+        // Using this context results in automatically reloading the app to display an site in the account.
+        case `default`
+        // Sign in to an existing account.
+        // This is typically used when the app needs to get a new OAuth token because the current one becomes invalid.
+        case reAuthentication(TaggedManagedObjectID<WPAccount>)
+        // Connect a site to Jetpack or sign in to an already connected site.
+        case jetpackSite(accountEmail: String?)
+
+        func accountEmail(in context: NSManagedObjectContext) -> String? {
+            switch self {
+            case .default:
+                return nil
+            case let .reAuthentication(accountID):
+                return (try? context.existingObject(with: accountID))?.email
+            case let .jetpackSite(email):
+                return email
+            }
+        }
+    }
+
+    enum AuthenticationError: Error {
         case invalidCallbackURL
         case loginDenied(message: String)
         case obtainAccessToken
         case urlError(URLError)
         case parsing(DecodingError)
         case cancelled
-        case unknown(Swift.Error)
+        case unknown(Error)
+    }
+
+    enum SignInError: Error {
+        case authentication(AuthenticationError)
+        case fetchUser(Error)
+        case mismatchedEmail(expectedEmail: String)
+        case alreadySignedIn(signedInAccountEmail: String)
+        case loadingSites(Error)
+    }
+
+    private let coreDataStack: CoreDataStackSwift
+    private let authenticator: ((URL) throws(AuthenticationError) -> URL)?
+
+    init(
+        coreDataStack: CoreDataStackSwift = ContextManager.shared,
+        authenticator: ((URL) throws(AuthenticationError) -> URL)? = nil
+    ) {
+        self.coreDataStack = coreDataStack
+        self.authenticator = authenticator
     }
 
+    /// Sign in WP.com account.
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
     @MainActor
-    func signIn(from viewController: UINavigationController) async {
+    func signIn(from viewController: UIViewController, context: SignInContext) async -> TaggedManagedObjectID<WPAccount>? {
+        WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "start"])
```

Comment:
What's the idea behind passing `"stage"` as opposed to creating separate events? 

Tracks is a bit limited in terms of its support of properties. It's usually easier to work with separate events.

Reply:
I saw something like `step` in the current login flow. I guess I took it from there.